### ISSUE TITLE
tmpfs size mount option not correct when expressed in %

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#5086]
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

Debian 232-10 which includes  https://github.com/systemd/systemd/pull/4835
git master

### Used distribution

Debian unstable
Originally reported as https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851492

### In case of bug report: Expected behaviour you didn't see

Specifying the tmpfs size as percentage should work

### In case of bug report: Unexpected behaviour you saw

Systemd does not mount tmpfs file systems with the correct size when
the size is expressed in % of physical RAM in /etc/fstab.

I guess it does not read at all the % character, and the size is read in
bytes. size=5% or size=10% results into file systems with a 4 k
size (man mount says "The size is given in bytes, and rounded up  to
entire  pages").

When expressed in k, m, g, the size is correct.

### In case of bug report: Steps to reproduce the problem

Add an entry like
```
tmpfs   /test        tmpfs   size=10%          0  0
```
to /etc/fstab. The tmpfs size will be 4k

A git bisect suggests this is a regression introduced by 
https://github.com/systemd/systemd/pull/4835

**Comment by evverx** [comment]
Indeed, https://github.com/systemd/systemd/pull/4835/commits/d107589cd2767c95fb1c794d4f90832301a297f5
> core: turn on specifier expansion for more unit file settings
Let's permit specifier expansion at a numbre of additional fields, where
arbitrary strings might be passed where this might be useful one day. (Or at
least where there's no clear reason where it wouldn't make sense to have.)

The quick fix is:
```diff
--- a/src/fstab-generator/fstab-generator.c
+++ b/src/fstab-generator/fstab-generator.c
@@ -347,8 +347,15 @@ static int add_mount(
         if (r < 0)
                 return r;

-        if (!isempty(filtered) && !streq(filtered, "defaults"))
-                fprintf(f, "Options=%s\n", filtered);
+        if (!isempty(filtered) && !streq(filtered, "defaults")) {
+                _cleanup_free_ char *o = NULL;
+
+                o = strreplace(filtered, "%", "%%");
+                if (!o)
+                        return log_oom();
+
+                fprintf(f, "Options=%s\n", o);
+        }

         r = fflush_and_check(f);
         if (r < 0)
```

Another question: do we really need d107589cd2767c95fb?

**Comment by poettering** [comment]
I figure the fstab generator should indeed escape %...

**Comment by poettering** [comment]
Fix is waiting in #5255. Please have a look.
